### DISCUSSION TITLE
Firefox idle-inhibit

### DISCUSSION THREAD

**Discussion by jp7677** [labwc/labwc#1503] [Show and tell]
I've set up `swayidle` on my system for screen locking and suspending. When watching a video with `firefox`, my screen still gets locked. The only work around I found was to actively inhibit idleness, e.g. using the corresponding way bar button. After spending some time I found the reason why inhibiting idleness isn't working in `firefox`, I'm posting this here so others don't have to dig into this rabbit hole.

Bottom line is that this come down to `xdg-desktop-portal` configuration and an outstanding issue in `xdg-desktop-portal`, the solution is very likely https://github.com/flatpak/xdg-desktop-portal/pull/1255

So `firefox` has multiple lock (inhibit) mechanisms implemented in a fixed order. This can be observed with `MOZ_LOG="LinuxWakeLock:5" firefox` and starting a video in the browser. It tries the first way to lock, if locking fails, it tries the next mechanism until it finds a call that succeeds. Unfortunately it tries the `Gnome` way before attempting to use Wayland `idle-inhibit`. The Gnome way is to use a `dbus` interface to prevent screensaver etc to kick-in. Even when running labwc/a wlroots based compositor, that dbus interface is exposed by xdg-desktop-portal. Not in the `wlr` interfaces (configured in `wlr.portal`), which are only about screen casting, but in the `gtk` fallback interfaces. 

According to the xdg-desktop-portal manual (`man portals.conf`), disabling an interface is possible using `none` as interface value. So the solution should be to create a `.config/xdg-desktop-portal/wlroots-portals.conf` file containing

```
[preferred]
default=gtk
org.freedesktop.impl.portal.ScreenCast=wlr
org.freedesktop.impl.portal.Screenshot=wlr
org.freedesktop.impl.portal.Inhibit=none
```

This is just a copy from `/usr/share/xdg-desktop-portal/wlroots-portals.conf` with additionally disabling the inhibit interface. Try `/usr/libexec/xdg-desktop-portal -v` to validate.

This should work, but it doesn't :( The reason is that the `none` specifier is broken in `xdg-desktop-portal`. An issue has been made here https://github.com/flatpak/xdg-desktop-portal/issues/1254 and there is even an open pull request https://github.com/flatpak/xdg-desktop-portal/pull/1255 . 

Whatever I tried on my machine, the current fallback mechanism in xdg-desktop-portal  always found a way to expose the gtk inhibit interface. So unfortunately this needs to get fixed in xdg-desktop-portal first. Furthermore I think the default `wlroots-portals.conf` should also contain `org.freedesktop.impl.portal.Inhibit=none` since exposing that interface on a wlroots based compositor is imho useless.

Anyway, long story, thanks for reading. I hope this PR gets merged in the near future and at least I learned a lot about the complexity of `idle-inhibit`.

Edit: The `xdg-portal-desktop` behavior has been fixed in https://github.com/flatpak/xdg-desktop-portal/releases/tag/1.18.4

PS: All file locations are for a standard Fedora installation. Other distributions might use different locations.

**Comment by Consolatis** [comment]
Nice research! Thanks for sharing,

Edit:
Note that setting `XDG_CURRENT_DESKTOP=labwc` may in turn break other applications that check for `wlroots` in that environment variable. One could add a `wlroots-portals.conf` instead and keep the env var set to `wlroots`.

Edit 2:
Another option could be to implement the `org.freedesktop.impl.portal.Inhibit` interface in the wlr portal implementation so it transparently uses the wayland idle protocol. I didn't verify if the protocols are compatible in all aspects.

**Reply by jp7677** [reply]
> Edit: Note that setting `XDG_CURRENT_DESKTOP=labwc` may in turn break other applications that check for `wlroots` in that environment variable. One could add a `wlroots-portals.conf` instead and keep the env var set to `wlroots`.

Yeah, that's better, thx! I've update my original post.

**Reply by Consolatis** [reply]
> Additionally XDG_CURRENT_DESKTOP=wlroots needs to be set in labwc's environment file, for xdg-desktop-portal to pick up the correct configuration file.

That should not be required, its is set automatically to `wlroots` to allow for screen capture with OBS and friends to work out of the box.

**Reply by jp7677** [reply]
ah right, it used to be there when I set up my environment, but indeed not anymore ( https://github.com/labwc/labwc/commit/5eb769d6d65c3b1474941606d178e6119c93b1d6 )

**Reply by jp7677** [reply]
> Edit 2: Another option could be to implement the `org.freedesktop.impl.portal.Inhibit` interface in the wlr portal implementation so it transparently uses the wayland idle protocol. I didn't verify if the protocols are compatible in all aspects.

For reference: https://github.com/emersion/xdg-desktop-portal-wlr/issues/297

**Comment by jp7677** [comment]
Good news, `xdg-desktop-portal` version 1.18.4 (https://github.com/flatpak/xdg-desktop-portal/releases/tag/1.18.4) contains the fixes for the issue I had mentioned above, so just creating a `wlroots-portals.conf` like described above works now.

**Reply by ammgws** [reply]
Since your workaround sets `default=gtk`, does that mean you need to keep xdg-desktop-portal-gtk (or whatever the equivalent is on your distro) installed even though it is overridden in the config to `wlr`/`none`?

**Reply by jp7677** [reply]
Yes. 
Note that Labwc nowadays ships its own config https://github.com/labwc/labwc/blob/master/data/labwc-portals.conf which defaults to wlr. But even if you are using that one, you should have `xdg-desktop-portal-gtk` installed since wlr defaults to gtk, see https://github.com/emersion/xdg-desktop-portal-wlr/blob/master/contrib/wlroots-portals.conf
